### ISSUE TITLE
Question: Possibility of override movement logic?

### ISSUE THREAD

**Issue by Anabolkick** [crashkonijn/GOAP#64]
Hi! I'm sorry if I'm asking a stupid question. As I read in the FAQ: "The graph calculates distance between two actions, adding that cost automatically. Actions don't need to be aware of each other, or their relative position in the graph, making them simpler." Where would it be possible to redefine the logic of moving to the target and calculating the cost of this movement? As I understand it, now everything is based on lifetime logic, and I would like to adapt it to turn based grid logic.

P.s. I found IAgentMover on Life Cycles of Agent (https://goap.crashkonijn.com/general/lifecycle) but don`t see this interface inside project(

**Comment by crashkonijn** [comment]
Hi @Anabolkick,

Thanks for having a go at this project!

Right now changing the calculation in the resolver isn't possible. I don't think it would be easy to change this, but I will think about this.

You're right that right now it's more geared towards running in realtime. I think it should be fairly simple for me to adjust the logic so it could be run turn-based as well.

Are there specific requirements that you'd need for your game? Like running a single agent at the time?

**Comment by Anabolkick** [comment]
Thank you so much for such a quick response, it was really unexpectedly quick. The fight gameplay involves logic, what player do some actions in one turn, then the turn goes to the enemies(agents). They make actions (not really important at the same time or one after another, although one after another is a little more preferable, because on this case enemy can consider position of his allies). 
And the peculiarity is that the opponent must determine what is better, to move a couple of cells and make a short-range attack or do some other action on current position.
The fight gameplay is similar to Trials of Fire or Battle Brothers.
Perhaps I wrote too much unnecessary information, but it's to make it easier to understand the specifics)
Thanks again for your attention!

**Comment by crashkonijn** [comment]
Haha I try to be as responsive as possible!

I will have a look at trying to implement this on Monday :)

**Comment by Anabolkick** [comment]
Cool! Thank you!

**Comment by crashkonijn** [comment]
FYI: I'm working on a solution but I figured this also requires a proper demo scene to test it. So it's a bit more work than expected.

I do believe I can get it working!

**Comment by crashkonijn** [comment]
Hi there!

I wanted to give you an update. I did get turns somewhat working, although it's not quite how I'd like it to work yet.

Tomorrow I'm getting the keys to my new house, so I wont be able to continue until next Monday.

Like I said, it's taking a bit longer than I expected but I want to do it properly :)

![goap_turns](https://github.com/crashkonijn/GOAP/assets/5364802/45ccea99-4a54-46e4-940c-13e21adbb41e)

**Comment by Anabolkick** [comment]
Oooh, congratulations on the housewarming! Thanks so much, I'll look forward to it) Thanks again)

**Comment by crashkonijn** [comment]
Hi there!

I just pushed the branch I was working on, with the demo scene as I have shown before. Our new house is taking too much time and I just can't find the energy to work on this at the moment.

I will continue with this, but it's probably going to take a while.

So what the demo scene already does:
- Example of turn based handling
- Example of limiting x movement per turn, completing paths over multiple turns

What it doesn't yet do:
- Optimized handling of the agents, each is still called each frame even when it's not their turn

Hopefully this is enough to get you going (for now). I will keep you posted!
-

**Comment by crashkonijn** [comment]
Hey @Anabolkick!

Your question 1.5 years ago got me thinking about how to do this properly, which led me down the rabbit hole of introducing the concepts of [controllers](https://goap.crashkonijn.com/goap-v3/general/controllers) into v3.

With v3 now almost being ready for release I finally found some time to implement the example that I made into v3 using the new `ManualController`.

This controller assumes that when you request a new goal to be resolved that you want it immediately and resolves it blocking the main thread. This should be perfect for turn based games!

All in all this took me way longer than I would've anticipated, but here it is: #298 🎉
